File: action_ws/src/my_action_pkg/src/spi_comm_server.cpp
// SPDX-License-Identifier: BSD-3-Clause

/**
 * @file spi_comm_server.cpp
 * @brief SPI 通信 Action Server (C++ 版本)
 * 
 * 功能说明:
 *   - 实现 ROS Action 接口，接收客户端发送的 SPI 通信请求
 *   - 使用 spidev 库通过 SPI 总线与设备通信
 *   - 支持写入数据并读取回环数据
 *   - 支持参数配置（SPI 总线、设备地址等）
 * 
 * 硬件连接:
 *   Jetson Nano SPI1 (/dev/spidev0.0):
 *     - 引脚 19 (GPIO 16/MOSI1): MOSI 主出从入
 *     - 引脚 21 (GPIO 17/MISO1): MISO 主入从出
 *     - 引脚 23 (GPIO 18/SCLK1): SCLK 时钟
 *     - 引脚 24 (GPIO 19/CE0): CE0 片选 0
 *     - 引脚 26 (GPIO 20/CE1): CE1 片选 1
 * 
 * 使用方法:
 *   source install/setup.bash
 *   rosrun my_action_pkg spi_comm_server_cpp
 * 
 * 作者: Jetson Nano
 * 日期: 2026-04-01
 */

#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include <my_action_pkg/SPICommAction.h>

#include <fcntl.h>
#include <linux/spi/spidev.h>
#include <sys/ioctl.h>
#include <unistd.h>
#include <cstring>
#include <cstdlib>

class SPICommServer {
private:
    ros::NodeHandle nh_;
    actionlib::SimpleActionServer<my_action_pkg::SPICommAction> server_;
    
    int spi_fd_;
    bool connected_;
    std::string enable_echo_;
    
public:
    SPICommServer() : 
        server_(nh_, "spi_comm", boost::bind(&SPICommServer::execute, this, _1), false),
        spi_fd_(-1),
        connected_(false)
    {
        nh_.param("enable_echo", enable_echo_, std::string("true"));
        
        ROS_INFO("Initializing SPI communication Action Server...");
        
        server_.start();
        
        ROS_INFO("SPI communication Action Server started");
        ROS_INFO("   Action name: spi_comm");
    }
    
    ~SPICommServer()
    {
        if (spi_fd_ >= 0) {
            close(spi_fd_);
            ROS_INFO("SPI bus closed");
        }
    }
    
    bool initSPI(uint8_t spi_bus, uint8_t device_address)
    {
        char device_path[32];
        snprintf(device_path, sizeof(device_path), "/dev/spidev%d.%d", spi_bus, device_address);
        
        if (spi_fd_ >= 0) {
            close(spi_fd_);
        }
        
        spi_fd_ = open(device_path, O_RDWR);
        if (spi_fd_ < 0) {
            ROS_ERROR("Failed to open SPI device: %s (Error: %s)", device_path, strerror(errno));
            connected_ = false;
            return false;
        }
        
        uint8_t mode = SPI_MODE_0;
        uint8_t bits = 8;
        uint32_t speed = 1000000;
        
        if (ioctl(spi_fd_, SPI_IOC_WR_MODE, &mode) < 0) {
            ROS_ERROR("Failed to set SPI mode");
            close(spi_fd_);
            connected_ = false;
            return false;
        }
        
        if (ioctl(spi_fd_, SPI_IOC_WR_BITS_PER_WORD, &bits) < 0) {
            ROS_ERROR("Failed to set SPI bits per word");
            close(spi_fd_);
            connected_ = false;
            return false;
        }
        
        if (ioctl(spi_fd_, SPI_IOC_WR_MAX_SPEED_HZ, &speed) < 0) {
            ROS_ERROR("Failed to set SPI clock speed");
            close(spi_fd_);
            connected_ = false;
            return false;
        }
        
        connected_ = true;
        ROS_INFO("SPI device opened successfully: %s", device_path);
        
        return true;
    }
    
    void execute(const my_action_pkg::SPICommGoalConstPtr &goal)
    {
        uint8_t spi_bus = goal->spi_bus;
        uint8_t device_address = goal->device_address;
        uint8_t write_data = goal->write_data;
        bool read_after_write = goal->read_after_write;
        
        ROS_INFO("Received Goal:");
        ROS_INFO("   SPI bus: %d", spi_bus);
        ROS_INFO("   Device address: %d", device_address);
        ROS_INFO("   Write data: 0x%02X (%d)", write_data, write_data);
        ROS_INFO("   Read after write: %s", read_after_write ? "true" : "false");
        
        if (!connected_) {
            if (!initSPI(spi_bus, device_address)) {
                my_action_pkg::SPICommResult result;
                result.received_data = 0;
                result.success = false;
                server_.setAborted(result);
                return;
            }
        }
        
        if (!connected_) {
            ROS_ERROR("SPI not connected, cannot execute task");
            my_action_pkg::SPICommResult result;
            result.received_data = 0;
            result.success = false;
            server_.setAborted(result);
            return;
        }
        
        my_action_pkg::SPICommFeedback feedback;
        feedback.status.data = "Writing data 0x" + std::to_string(write_data);
        feedback.current_data = write_data;
        server_.publishFeedback(feedback);
        
        try {
            if (enable_echo_ == "true") {
                ROS_INFO("Writing data: 0x%02X", write_data);
            }
            
            uint8_t received_data = 0;
            struct spi_ioc_transfer tr = {};
            tr.tx_buf = (unsigned long)&write_data;
            tr.rx_buf = (unsigned long)&received_data;
            tr.len = 1;
            tr.speed_hz = 1000000;
            tr.bits_per_word = 8;
            
            if (ioctl(spi_fd_, SPI_IOC_MESSAGE(1), &tr) < 0) {
                ROS_ERROR("SPI communication error: %s", strerror(errno));
                my_action_pkg::SPICommResult result;
                result.received_data = 0;
                result.success = false;
                server_.setAborted(result);
                return;
            }
            
            if (read_after_write) {
                if (enable_echo_ == "true") {
                    ROS_INFO("Read data: 0x%02X (%d)", received_data, received_data);
                }
                
                feedback.status.data = "Read data: 0x" + std::to_string(received_data);
                feedback.current_data = received_data;
                server_.publishFeedback(feedback);
                
                my_action_pkg::SPICommResult result;
                result.received_data = received_data;
                result.success = true;
                
                server_.setSucceeded(result);
                ROS_INFO("SPI communication succeeded");
            } else {
                if (enable_echo_ == "true") {
                    ROS_INFO("Data written successfully, no read");
                }
                
                my_action_pkg::SPICommResult result;
                result.received_data = 0;
                result.success = true;
                
                server_.setSucceeded(result);
                ROS_INFO("SPI write succeeded");
            }
            
        } catch (const std::exception& e) {
            ROS_ERROR("SPI communication error: %s", e.what());
            my_action_pkg::SPICommResult result;
            result.received_data = 0;
            result.success = false;
            server_.setAborted(result);
        }
    }
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "spi_comm_server_cpp");
    
    SPICommServer server;
    
    ROS_INFO("Action Server running, waiting for Goal requests...");
    
    ros::spin();
    
    return 0;
}

File: action_ws/install/lib/my_action_pkg/spi_comm_server.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
========================================
SPI 通信 Action Server
========================================
功能说明:
  - 实现 ROS Action 接口，接收客户端发送的 SPI 通信请求
  - 使用 spidev 库通过 SPI 总线与设备通信
  - 支持写入数据并读取回环数据
  - 支持参数配置（SPI 总线、设备地址等）
  
硬件连接:
  Jetson Nano SPI0:
    - 引脚 19 (GPIO 10/MOSI1): MOSI 主出从入
    - 引脚 21 (GPIO 9/MISO1): MISO 主入从出
    - 引脚 23 (GPIO 11/SCLK1): SCLK 时钟
    - 引脚 24 (GPIO 8/CE0): CE0 片选 0
    - 引脚 26 (GPIO 7/CE1): CE1 片选 1
    - 设备路径: /dev/spidev0.0 (CE0), /dev/spidev0.1 (CE1)
  
使用方法:
  source install/setup.bash
  sudo rosrun my_action_pkg spi_comm_server.py
  
作者: Jetson Nano
日期: 2026-03-31
"""

import rospy
import actionlib
import spidev
from my_action_pkg.msg import SPICommAction, SPICommFeedback, SPICommResult
from std_msgs.msg import String


class SPICommServer:
    """
    SPI 通信 Action Server 类
    
    职责:
      1. 初始化 ROS 节点和 SPI 总线
      2. 创建 Action Server 接收客户端请求
      3. 执行 SPI 写入和读取操作
      4. 返回操作结果给客户端
    """
    
    def __init__(self):
        """
        构造函数 - 初始化服务器
        
        功能:
          1. 初始化 ROS 节点
          2. 从参数服务器读取配置参数
          3. 初始化 SPI 总线
          4. 创建并启动 Action Server
        """
        
        rospy.init_node('spi_comm_server')
        rospy.loginfo("正在初始化 SPI 通信 Action Server...")
        
        self.enable_echo = rospy.get_param('~enable_echo', True)
        
        self.spi = None
        self.connected = False
        
        rospy.loginfo(f"✅ SPI 通信 Action Server 已启动")
        rospy.loginfo(f"   Action 名称: spi_comm")
        
        self.server = actionlib.SimpleActionServer(
            'spi_comm', 
            SPICommAction, 
            self.execute, 
            False
        )
        self.server.start()
    
    def init_spi(self, spi_bus, device_address):
        """
        初始化 SPI 总线
        
        参数:
          spi_bus: SPI 总线编号 (0, 1, 2)
          device_address: 设备地址/片选编号 (0, 1)
        """
        try:
            if self.spi:
                self.spi.close()
            
            self.spi = spidev.SpiDev()
            self.spi.open(spi_bus, device_address)
            
            self.spi.max_speed_hz = 1000000
            self.spi.mode = 0
            self.spi.bits_per_word = 8
            
            self.connected = True
            rospy.loginfo(f"✅ SPI 已成功打开: /dev/spidev{spi_bus}.{device_address}")
            
        except FileNotFoundError:
            self.connected = False
            rospy.logerr(f'❌ SPI 设备不存在: /dev/spidev{spi_bus}.{device_address}')
            
        except PermissionError:
            self.connected = False
            rospy.logerr(f'❌ 权限不足: 无法访问 /dev/spidev{spi_bus}.{device_address}')
            rospy.logerr('   提示: 请使用 sudo 运行')
            
        except Exception as e:
            self.connected = False
            rospy.logerr(f'❌ SPI 初始化错误: {e}')
    
    def execute(self, goal):
        """
        执行 Goal 的回调函数
        
        参数:
          goal: 客户端发送的目标，包含 SPI 通信参数
        """
        spi_bus = goal.spi_bus
        device_address = goal.device_address
        write_data = goal.write_data
        read_after_write = goal.read_after_write
        
        rospy.loginfo(f"📥 收到 Goal:")
        rospy.loginfo(f"   SPI 总线: {spi_bus}")
        rospy.loginfo(f"   设备地址: {device_address}")
        rospy.loginfo(f"   写入数据: 0x{write_data:02X} ({write_data})")
        rospy.loginfo(f"   写后读取: {read_after_write}")
        
        if not self.connected:
            self.init_spi(spi_bus, device_address)
        
        if not self.connected:
            rospy.logerr('❌ SPI 未连接，无法执行任务')
            result = SPICommResult()
            result.received_data = 0
            result.success = False
            self.server.set_aborted(result)
            return
        
        feedback = SPICommFeedback()
        feedback.status = String(data=f"正在写入数据 0x{write_data:02X}")
        feedback.current_data = write_data
        self.server.publish_feedback(feedback)
        
        try:
            if self.enable_echo:
                rospy.loginfo(f"📤 正在写入数据: 0x{write_data:02X}")
            
            received_data = self.spi.xfer2([write_data])
            
            if read_after_write:
                if self.enable_echo:
                    rospy.loginfo(f"📥 读取数据: 0x{received_data[0]:02X} ({received_data[0]})")
                
                feedback = SPICommFeedback()
                feedback.status = String(data=f"读取数据: 0x{received_data[0]:02X}")
                feedback.current_data = received_data[0]
                self.server.publish_feedback(feedback)
                
                result = SPICommResult()
                result.received_data = received_data[0]
                result.success = True
                
                self.server.set_succeeded(result)
                rospy.loginfo("✅ SPI 通信成功")
            else:
                if self.enable_echo:
                    rospy.loginfo(f"✅ 数据写入完成，不读取")
                
                result = SPICommResult()
                result.received_data = 0
                result.success = True
                
                self.server.set_succeeded(result)
                rospy.loginfo("✅ SPI 写入成功")
                
        except Exception as e:
            rospy.logerr(f'❌ SPI 通信错误: {e}')
            result = SPICommResult()
            result.received_data = 0
            result.success = False
            self.server.set_aborted(result)
    
    def cleanup(self):
        """
        清理资源
        """
        if self.spi:
            try:
                self.spi.close()
                rospy.loginfo('SPI 总线已关闭')
            except Exception as e:
                rospy.logerr(f'关闭 SPI 总线时出错: {e}')


if __name__ == '__main__':
    server = None
    try:
        server = SPICommServer()
        rospy.loginfo("🔄 Action Server 正在运行，等待 Goal 请求...")
        rospy.spin()
        
    except rospy.ROSInterruptException:
        rospy.loginfo('📛 收到中断信号，正在关闭...')
        
    except Exception as e:
        rospy.logerr(f'❌ 节点运行错误: {e}')
        
    finally:
        if server:
            server.cleanup()